snapshot/snapshot.sh

```diff
@@ -0,0 +1,353 @@
+    rm -rf $IMAGE_DIR >& /dev/null
+    rm $SNAP_DOCKER_FILE >& /dev/null
+
+}
+
+function print_help() {
+    echo "usage: $0 deployment-name"
+    echo ""
+    echo "Create snapshot of a deployment pod and apply if needed. "
+    echo "Environment variable KONTAIN_SNAPSHOT_REGISTRY must be set prior to calling script in the format"
+    echo " registry/account "
+    echo " For example, export KONTAIN_SNAPSHOT_REGISTRY=docker.io/kontainapp."
+    echo "If repository is private, make sure to login prior to calling the script"
+    echo ""
+    echo "Options:"
+    echo "  deployment-name Name of the deployment as printed by command \
+                kubectl get deployments"
+    exit 1
+}
+
+function load_tools() {
+    # Get yq
+    YQ=$DIR/yq
+    if [ ! -f $YQ ] && [ ! -x $YQ ]; then 
+        curl -o $YQ -L https://github.com/mikefarah/yq/releases/download/v4.28.1/yq_linux_amd64
+        chmod +x YQ
+    fi
+}
+
+function add_envs()
+{    
+    filename=$1
+    # Add KM_MGTDIR env var to the container
+    # Add KM_MGTPIPE env var to allow old km's to work.
+    declare -A kontain_envs
+    kontain_envs[KM_MGTDIR]=$SNAPDIR_CONTAINER
+    kontain_envs[KM_MGTPIPE]=$SNAPDIR_CONTAINER/kmpipe.$$
+    kontain_envs[SNAPDIR_HOST]=$SNAPDIR_HOST
+    kontain_envs[SNAPDIR_CONTAINER]=$SNAPDIR_CONTAINER
+    
+    if [ $# -eq 2 ]; then
+        # declare a local **reference variable** (hence `-n`) named `data_ref`
+        # which is a reference to the value stored in the first parameter
+        # passed in
+        local -n data_ref="$2"
+        for k in "${!data_ref[@]}"; do
+            kontain_envs["$k"]="${data_ref["$k"]}"
+        done
+    fi
+
+    # check for every key and add or update it 
+    for i in "${!kontain_envs[@]}"
+    do
+        has_key=$(key=$i $YQ '.spec.containers[].env[]|select(.name == strenv(key)).name' $filename)
+        if [ "$has_key" = "$i" ]; then 
+            key=$i value="${kontain_envs[$i]}" $YQ -i '(.spec.containers[].env[]|select(.name == strenv(key)).value)|=strenv(value)' $filename
+        else
+            key=$i value="${kontain_envs[$i]}" $YQ -i '.spec.containers[].env = [{"name": strenv(key), "value": strenv(value)}] + .spec.containers[].env' $filename
+        fi
+    done
+}
+
+function set_node_selector() {
+    filename=$1
+
+    #delete nodeName if any
+    $YQ -i 'del(.spec.nodeName)' $filename
+
+    #add node selector
+    host=$NODE_NAME $YQ -i '.spec.nodeSelector += {"kubernetes.io/hostname": strenv(host)}' $filename
+
+}
+
+function get_pod() {
+    echo -n "Reading deployment information....."
+
+    # find a pod that belongs to the deployment and get its yaml
+
+    selector=$(kubectl get deployment $deployment_name -o wide | tail -n 1 | awk '{print $8}')
+
+    pod=$(kubectl get pods -l $selector | tail -1 | awk '{print $1}')
+
+    kubectl get pod $pod -o yaml > $POD_YAMLFILE
+
+    # read image the pod was created from
+    IMAGE_NAME=$($YQ '.spec.containers[].image' $POD_YAMLFILE)
```

Comment:
For now we should not attempt snapshots when the pod/deployment contains more than one container.
I haven't checked to see if this is excluded elsewhere.

Reply:
I wil check for number of containers and exit f more then one with a message